shell/platform/fuchsia/flutter/software_surface.cc

```diff
@@ -0,0 +1,404 @@
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to allocate collection: "
+                   << zx_status_get_string(allocate_status);
+    return false;
+  }
+  fuchsia::sysmem::BufferCollectionTokenSyncPtr scenic_token;
+  zx_status_t duplicate_status = local_token->Duplicate(
+      std::numeric_limits<uint32_t>::max(), scenic_token.NewRequest());
+  if (duplicate_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to duplicate collection token: "
+                   << zx_status_get_string(duplicate_status);
+    return false;
+  }
+  zx_status_t sync_status = local_token->Sync();
+  if (sync_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to sync collection token: "
+                   << zx_status_get_string(sync_status);
+    return false;
+  }
+
+  if (session_) {
+    session_->RegisterBufferCollection(buffer_id, std::move(scenic_token));
+    buffer_id_ = buffer_id;
+  } else {
+    fuchsia::ui::composition::BufferCollectionExportToken export_token;
+    zx_status_t token_create_status =
+        zx::eventpair::create(0, &export_token.value, &import_token_.value);
+    if (token_create_status != ZX_OK) {
+      FML_LOG(ERROR)
+          << "SoftwareSurface: Failed to create flatland export token: "
+          << zx_status_get_string(token_create_status);
+      return false;
+    }
+
+    fuchsia::ui::composition::RegisterBufferCollectionArgs args;
+    args.set_export_token(std::move(export_token));
+    args.set_buffer_collection_token(std::move(scenic_token));
+    args.set_usage(
+        fuchsia::ui::composition::RegisterBufferCollectionUsage::DEFAULT);
+    flatland_allocator->RegisterBufferCollection(
+        std::move(args),
+        [](fuchsia::ui::composition::Allocator_RegisterBufferCollection_Result
+               result) {
+          if (result.is_err()) {
+            FML_LOG(ERROR)
+                << "RegisterBufferCollection call to Scenic Allocator failed";
+          }
+        });
+  }
+
+  fuchsia::sysmem::BufferCollectionSyncPtr buffer_collection;
+  zx_status_t bind_status = sysmem_allocator->BindSharedCollection(
+      std::move(local_token), buffer_collection.NewRequest());
+  if (bind_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to bind collection token: "
+                   << zx_status_get_string(bind_status);
+    return false;
+  }
+
+  fuchsia::sysmem::BufferCollectionConstraints constraints;
+  constraints.min_buffer_count = 1;
+  constraints.usage.cpu =
+      fuchsia::sysmem::cpuUsageWrite | fuchsia::sysmem::cpuUsageWriteOften;
+  constraints.has_buffer_memory_constraints = true;
+  constraints.buffer_memory_constraints.min_size_bytes = 0;
+  constraints.buffer_memory_constraints.max_size_bytes = 0xffffffff;
+  constraints.buffer_memory_constraints.physically_contiguous_required = false;
+  constraints.buffer_memory_constraints.secure_required = false;
+  constraints.buffer_memory_constraints.ram_domain_supported = true;
```

Comment:
If you allow RAM domain, you need to flush after writing to it. See https://cs.opensource.google/fuchsia/fuchsia/+/master:src/ui/scenic/lib/gfx/tests/scenic_pixeltest.cc;drc=5796f5829de3eb33efc3c51dd739eded4f9abf9b;l=1408

Reply:
The ARM Mali GPU driver (used by scenic on some ARM platforms) requires RAM domain. If you don't flush after writing, you can get tearing and corruption when reading from the GPU. If you read from the buffer and another client wrote to it you also need to invalidate before the read, but that doesn't seem to be necessary here.

Reply:
Done